core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala

```diff
@@ -52,43 +51,84 @@ class TransactionMarkerRequestCompletionHandler(transactionMarkerChannel: Transa
+          case None =>
+            info(s"Transaction topic partition for $transactionalId may likely has emigrated, as the corresponding metadata do not exist in the cache" +
+              s"any more; cancel sending transaction markers $txnMarker to the brokers")
+
+            // txn topic partition has likely emigrated, just cancel it from the purgatory
+            txnMarkerChannelManager.removeMarkersForTxnId(transactionalId)
+
+          case Some(epochAndMetadata) =>
+            val txnMetadata = epochAndMetadata.transactionMetadata
+
+            if (epochAndMetadata.coordinatorEpoch != txnMarker.coordinatorEpoch) {
+              // coordinator epoch has changed, just cancel it from the purgatory
+              info(s"Transaction coordinator epoch for $transactionalId has changed from ${txnMarker.coordinatorEpoch} to " +
+                s"${epochAndMetadata.coordinatorEpoch}; cancel sending transaction markers $txnMarker to the brokers")
+
+              txnMarkerChannelManager.removeMarkersForTxnId(transactionalId)
+            }
+
+            txnMetadata synchronized {
+              for ((topicPartition: TopicPartition, error: Errors) <- errors) {
+                error match {
+                  case Errors.NONE =>
+
+                    txnMetadata.removePartition(topicPartition)
+
+                  case Errors.CORRUPT_MESSAGE |
+                       Errors.MESSAGE_TOO_LARGE |
+                       Errors.RECORD_LIST_TOO_LARGE |
+                       Errors.INVALID_REQUIRED_ACKS => // these are all unexpected and fatal errors
+
+                    throw new IllegalStateException(s"Received fatal error ${error.exceptionName} while sending txn marker for $transactionalId")
+
+                  case Errors.UNKNOWN_TOPIC_OR_PARTITION |
+                       Errors.NOT_LEADER_FOR_PARTITION |
+                       Errors.NOT_ENOUGH_REPLICAS |
+                       Errors.NOT_ENOUGH_REPLICAS_AFTER_APPEND => // these are retriable errors
+
+                    info(s"Sending $transactionalId's transaction marker for partition $topicPartition has failed with error ${error.exceptionName}, retrying " +
+                      s"with current coordinator epoch ${epochAndMetadata.coordinatorEpoch}")
+
+                    retryPartitions += topicPartition
+
+                  case Errors.INVALID_PRODUCER_EPOCH |
+                       Errors.TRANSACTION_COORDINATOR_FENCED => // producer or coordinator epoch has changed, this txn can now be ignored
+
+                    info(s"Sending $transactionalId's transaction marker for partition $topicPartition has permanently failed with error ${error.exceptionName} " +
+                      s"with the current coordinator epoch ${epochAndMetadata.coordinatorEpoch}; cancel sending any more transaction markers $txnMarker to the brokers")
+
+                    txnMarkerChannelManager.removeMarkersForTxnId(transactionalId)
```

Comment:
Should we return here? Otherwise, the call to `completeSendMarkersForTxnId` will be invoked below. Not sure it's a problem, but seems odd.

Also, I'm not sure I fully understand the chain of operations that this should trigger. It seems that `removeMarkersForTxnId` cancels the `DelayedTxnMarker`, which means its callback won't get invoked. But what happens to the state of the `TransactionMetadata`? It seems like it will just remain indefinitely in one of the Prepare states. For CoordinatorFenced, that seems OK; we rely on the partition to ultimately be evicted. How about ProducerFenced? I'm actually having a tough time imagining a scenario where we would hit that error. The only one that comes to mind is if the coordinator has become a zombie, in which case we should get blocked by the coordinator epoch. Are there any others? In any case, I think it would be helpful to add some comments here clarifying the scenario and the handling expectation.